File: backtest/optimizer/optimizer.go
package optimizer

import (
	"context"

	"quantmesh/backtest"
	"quantmesh/exchange"
)

// Range 浮点参數範圍 (min, max, step)
type Range struct {
	Min  float64 `json:"min"`
	Max  float64 `json:"max"`
	Step float64 `json:"step"`
}

// IntRange 整數参數範圍
type IntRange struct {
	Min  int `json:"min"`
	Max  int `json:"max"`
	Step int `json:"step"`
}

// OptimSearchSpace 参數搜索空间
type OptimSearchSpace struct {
	PriceLowRange   Range   `json:"price_low_range"`
	PriceHighRange  Range   `json:"price_high_range"`
	GridCountRange  IntRange `json:"grid_count_range"`
	OrderQtyRange   Range   `json:"order_qty_range"`
}

// OptimConfig 优化配置
type OptimConfig struct {
	Method        string  `json:"method"`         // "grid", "bayesian", "genetic"
	Lambda        float64 `json:"lambda"`         // 风險权重 0.3-0.7
	MaxIterations int     `json:"max_iterations"` // 最大迭代次數
	Tolerance     float64 `json:"tolerance"`     // 收敛容差
	Parallelism   int     `json:"parallelism"`   // 並行度，0 表示 NumCPU
}

// Optimizer 优化器接口
type Optimizer interface {
	// Run 執行优化，返回最优結果；ctx 用於取消
	Run(ctx context.Context, symbol string, candles []*exchange.Candle, space OptimSearchSpace, config OptimConfig, initialCapital float64) (*OptimResult, error)
}

// BacktestRunner 單次回测執行器，供各优化器調用
func BacktestRunner(symbol string, candles []*exchange.Candle, params backtest.GridBacktestParams, initialCapital float64) (*backtest.BacktestResult, error) {
	return backtest.RunGridBacktest(symbol, candles, params, initialCapital)
}

// ParamsFromSpace 從搜索空间生成單组回测参數（用於固定 FeeRate 等）
func ParamsFromSpace(priceLow, priceHigh float64, gridCount int, orderQty, totalCapital, feeRate, slippage float64) backtest.GridBacktestParams {
	if feeRate <= 0 {
		feeRate = 0.0004
	}
	if slippage <= 0 {
		slippage = 0.0003
	}
	return backtest.GridBacktestParams{
		PriceLow:       priceLow,
		PriceHigh:      priceHigh,
		GridCount:      gridCount,
		OrderQuantity:  orderQty,
		TotalCapital:   totalCapital,
		FeeRate:        feeRate,
		SlippageRatio:  slippage,
	}
}

// DefaultOptimConfig 返回默认优化配置
func DefaultOptimConfig() OptimConfig {
	return OptimConfig{
		Method:        "grid",
		Lambda:        0.5,
		MaxIterations: 100,
		Tolerance:     1e-4,
		Parallelism:   0,
	}
}

// ValidateSearchSpace 校驗搜索空间合法性
func ValidateSearchSpace(space OptimSearchSpace) error {
	if space.PriceLowRange.Min >= space.PriceLowRange.Max {
		return errInvalidRange
	}
	if space.PriceHighRange.Min >= space.PriceHighRange.Max {
		return errInvalidRange
	}
	if space.PriceLowRange.Step <= 0 || space.PriceHighRange.Step <= 0 {
		return errInvalidRange
	}
	if space.GridCountRange.Min <= 0 || space.GridCountRange.Max < space.GridCountRange.Min {
		return errInvalidRange
	}
	if space.OrderQtyRange.Min <= 0 || space.OrderQtyRange.Max < space.OrderQtyRange.Min {
		return errInvalidRange
	}
	return nil
}

File: webui/src/services/optimizer.ts
import { fetchWithAuth } from './api'

const API_BASE_URL = `${window.location.origin}/api`

// 参數範圍
export interface Range {
  min: number
  max: number
  step: number
}

export interface IntRange {
  min: number
  max: number
  step: number
}

// 搜索空间
export interface OptimSearchSpace {
  price_low_range: Range
  price_high_range: Range
  grid_count_range: IntRange
  order_qty_range: Range
}

// 优化配置
export interface OptimConfig {
  method: 'grid' | 'bayesian' | 'genetic'
  lambda: number
  max_iterations: number
  tolerance: number
  parallelism: number
}

// 优化任務状態
export interface OptimizerTaskStatus {
  success: boolean
  task_id: string
  status: string
  progress: number
  error: string
  created_at: string
  updated_at: string
}

// 网格参數結果
export interface GridBacktestParams {
  price_low: number
  price_high: number
  grid_count: number
  order_quantity: number
  total_capital: number
  fee_rate: number
  slippage_ratio: number
}

// 性能指標
export interface Metrics {
  total_return: number
  annualized_return: number
  max_drawdown: number
  max_drawdown_duration: number
  volatility: number
  sharpe_ratio: number
  sortino_ratio: number
  calmar_ratio: number
  total_trades: number
  win_rate: number
  profit_factor: number
  avg_win: number
  avg_loss: number
  largest_win: number
  largest_loss: number
  max_consecutive_wins: number
  max_consecutive_losses: number
}

// 單组参數結果
export interface ParamResult {
  params: GridBacktestParams
  score: number
  metrics: Metrics
}

// 热力图數據
export interface HeatmapData {
  x_axis: (number | string)[]
  y_axis: (number | string)[]
  data: number[][]
}

// 优化結果
export interface OptimResult {
  best_params: GridBacktestParams
  best_score: number
  best_metrics: Metrics
  all_results: ParamResult[]
  heatmap_data: HeatmapData | null
  elapsed: number
  iterations: number
  method: string
}

// 啟动优化任務
export async function postOptimizerRun(params: {
  exchange?: string
  symbol: string
  interval: string
  start_time: string
  end_time: string
  initial_capital: number
  search_space: OptimSearchSpace
  config: OptimConfig
}): Promise<{ success: boolean; message: string; task_id: string }> {
  return fetchWithAuth(`${API_BASE_URL}/optimizer/run`, {
    method: 'POST',
    body: JSON.stringify(params),
  })
}

// 查詢任務状態
export async function getOptimizerStatus(id: string): Promise<OptimizerTaskStatus> {
  return fetchWithAuth(`${API_BASE_URL}/optimizer/status/${encodeURIComponent(id)}`)
}

// 獲取优化結果
export async function getOptimizerResult(id: string): Promise<{ success: boolean; status: string; result: OptimResult | null }> {
  return fetchWithAuth(`${API_BASE_URL}/optimizer/result/${encodeURIComponent(id)}`)
}

// 獲取交易對當前價格（用於自动初始化搜索空间）
export async function getOptimizerPrice(exchange: string, symbol: string): Promise<{ price: number; symbol: string; exchange: string }> {
  const params = new URLSearchParams({ exchange, symbol })
  return fetchWithAuth(`${API_BASE_URL}/optimizer/price?${params}`)
}

// 停止优化任務
export async function postOptimizerStop(id: string): Promise<{ success: boolean; message: string }> {
  return fetchWithAuth(`${API_BASE_URL}/optimizer/stop/${encodeURIComponent(id)}`, { method: 'POST' })
}
